### ISSUE TITLE
bug(cmd save): qri save doesn't detect changes in body

### ISSUE THREAD

**Issue by Arqu** [qri-io/qri#1383]
### Environment

# OS info
```
ProductName:    Mac OS X
ProductVersion: 10.15.2
BuildVersion:   19C57
```

# qri version
```
0.9.9-dev @ 47415190d37e5f25a8b3a3c60950467809ab2745
```

### Expected Behavior
```
mkdir population_bank_world
cd population_bank_world
qri init
cp ../body.csv ./body.csv
qri save
echo "Zimbabwe,ZWE,\"Population, total\",SP.POP.TOTL,3776681,3905034,4039201,4178726,4322861,4471177,4623351,4779827,4941906,5111337,5289303,5476982,5673911,5877726,6085074,6293875,6502569,6712827,6929664,7160023,7408624,7675591,7958241,8254747,8562249,8877489,9200149,9527203,9849125,10153852,10432421,10680995,10900502,11092766,11261744,11410714,11541217,11653242,11747072,11822719,11881477,11923914,11954290,11982224,12019912,12076699,12155491,12255922,12379549,12526968,12697723,12894316,13115131,13350356,13586681,13814629,14030390,14236745,14439018,," >> ./body.csv
qri save
```

Expected to produce:
```
for linked dataset [arqu/population_bank_world]

dataset saved: arqu/population_bank_world@/ipfs/QmVx438tG5FZLM2xLaz9YD57RPSVhDR2fFv8QfaDuGRaok
```
### Actual Behavior

```
for linked dataset [arqu/population_bank_world]

error saving: no changes
```

### Other info
- Dataset body taken from `b5/world_bank_population`

**Comment by Arqu** [comment]
So I'm re-opening this as I've managed to encounter this with an empty repo.
Here's a branch that logs out the following (`--log-all`) https://github.com/qri-io/qri/commit/217b8dbd99b8a84c8c09c59124468d737f26bcba

```
for linked dataset [hepek/wbp]

⠋ DiffDescriptions 0
DiffDescriptions 1
buildComponentChanges HD:9, BD:3
buildComponentChanges type:replace
buildComponentChanges initCompChanges
buildComponentChanges type:+
buildComponentChanges initCompChanges
buildComponentChanges type:+
buildComponentChanges initCompChanges
buildComponentChanges assumeBody NOT Changed
buildBodyChanges parentPath: ''
buildBodyChanges delta: ' '
buildBodyChanges delta: ' '
buildBodyChanges delta: ' '
DiffDescriptions comp:meta
DiffDescriptions comp:structure
DiffDescriptions comp:readme
DiffDescriptions comp:viz
DiffDescriptions comp:transform
DiffDescriptions comp:body
DiffDescriptions 2
DiffDescriptions 3
error saving: no changes
```

Attached are the repo, config and dataset.
[bugdrop.zip](https://github.com/qri-io/qri/files/4767104/bugdrop.zip)

Notable differences in setup:
- Using local registry

**Comment by dustmop** [comment]
Still unable to reproduce this. But looks at the logging output you added I'm noticing the following:

```
buildComponentChanges HD:9, BD:3
...
buildBodyChanges parentPath: ''
buildBodyChanges delta: ' '
buildBodyChanges delta: ' '
buildBodyChanges delta: ' '
```

This means that the bodyDeltas has 3 deltas in it, which is what gets returned from `deepdiff`. Iterating over those deltas, your code is printing the delta.Type, which is the `' '`. This means, of course, `DTContext` https://github.com/qri-io/deepdiff/blob/master/delta.go#L13, which is represented by a space character and is used when nothing is different between the first and second values passed to deepdiff. So for whatever reason, the bodys that are handed off to deepdiff result in a diff with no changes.

---

One other thing, just to prevent following any incorrect leads, this part of the log:

```
buildComponentChanges assumeBody NOT Changed
```

should actually be stated as:

```
buildComponentChanges NOT assumeBodyChanged
```

in other words, we shouldn't assume the body has changed, because we should to inspect the result of deepdiff to determine that fact. The `assumeBodyChanged` is only true when the body is too large to diff, which is *not* the case here.

**Comment by Arqu** [comment]
So got another update on this.
I can reliably reporduce this from a fresh repo.
If the initial body is:
```
one,two,3
four,five,6
```
And I update it to:
```
one,two,3
four,five,6
four,five,6
```
It doesn't detect the changes, however if I instead update it to:
```
one,two,3
four,five,6
four,five,7
```
It works  and saves correctly.

Now I'm not entirely sure to why it is happening (especially since we accurately detect the presence of a change on `qri status`, however my hunch would be that for status we only have a lightweight check like size/hash equality while the save path includes the differ to infer all the subesquent changes. 

Now the differ might fail on the above if the rows are the same as it doesn't have enough context bellow/above and figures it is "the same row" and thus doesn't detect a "real" change in the contents. If you change the value, it obviously detects the row diff and then works as expected.

@dustmop @b5 any thoughts on this?

**Comment by Arqu** [comment]
Just to add newer info, seems like if I do `qri save` any number of times it fails. However if I do `qri status` and then `qri save` it picks up.

**Comment by Arqu** [comment]
Think this is another instance of this: https://github.com/qri-io/qri/issues/1505

**Comment by dustmop** [comment]
Would deeply enjoy fixing this bug, but likely it will require a rewrite, or at least a large overhaul of deepdiff. Part of the problem is that deepdiff contains non-determinism: https://github.com/qri-io/deepdiff/pull/7, but even aside from that, it returns incorrect results in certain known cases.

The save path uses deepdiff for 2 purposes: 1) determining whether something in the body has changed, and 2) deriving a textual message that describes what the change is. As long as deepdiff has these bugs, the save operation is going to behave incorrectly.

**Comment by ramfox** [comment]
hey just adding my own notation of seeing this bug, but running `qri status` before running `qri save` **did not** allow me to save
